src/stream/src/common/table/state_table.rs

```diff
@@ -699,30 +705,40 @@ where
         let (_, output_indices) = find_columns_by_ids(&columns[..], &output_column_ids);
 
         // Get clean watermark PK index using the helper method
+        assert!(
+            table_catalog.clean_watermark_indices.len() <= 1,
+            "more than 1 watermark indices found: {:?}",
+            table_catalog.clean_watermark_indices
+        );
         let clean_watermark_index_in_pk: Option<i32> = table_catalog
             .get_clean_watermark_index_in_pk_compat()
             .map(|idx| idx as i32);
+        let clean_watermark_index_in_value = table_catalog
+            .get_clean_watermark_column_index_in_value()
+            .map(|idx| idx as i32);
 
         // Restore persisted table watermark.
-        let watermark_serde = if pk_indices.is_empty() {
-            None
-        } else {
-            // Use the watermark PK index from clean_watermark_index_in_pk
-            let pk_idx = clean_watermark_index_in_pk.unwrap_or(0) as usize;
-            Some(pk_serde.index(pk_idx))
-        };
+        let pair = watermark_serde_and_type(
```

Comment:
I think we can store the result to state table to be reused when committing watermarks.